tree/ntuple/v7/src/RNTupleDescriptor.cxx

```diff
@@ -1046,6 +1067,46 @@ ROOT::Experimental::Internal::RNTupleDescriptorBuilder::AddExtraTypeInfo(RExtraT
    return RResult<void>::Success();
 }
 
+ROOT::Experimental::Internal::RNTupleSerializer::StreamerInfoMap_t
+ROOT::Experimental::Internal::RNTupleDescriptorBuilder::BuildStreamerInfos() const
+{
+   RNTupleSerializer::StreamerInfoMap_t streamerInfoMap;
+   const auto &desc = GetDescriptor();
+
+   std::function<void(const RFieldDescriptor &)> fnWalkFieldTree;
+   fnWalkFieldTree = [&desc, &streamerInfoMap, &fnWalkFieldTree](const RFieldDescriptor &fieldDesc) {
+      if (fieldDesc.IsCustomClass()) {
+         // Add streamer info for this class to streamerInfoMap
+         auto cl = TClass::GetClass(fieldDesc.GetTypeName().c_str());
+         if (!cl) {
+            throw RException(R__FAIL(std::string("cannot get TClass for ") + fieldDesc.GetTypeName()));
+         }
+         auto streamerInfo = cl->GetStreamerInfo(fieldDesc.GetTypeVersion());
+         if (!streamerInfo) {
+            throw RException(R__FAIL(std::string("cannot get streamerInfo for ") + fieldDesc.GetTypeName()));
+         }
+         streamerInfoMap[streamerInfo->GetNumber()] = streamerInfo;
+      }
+
+      // Recursively traverse sub fields
+      for (const auto &subFieldDesc : desc.GetFieldIterable(fieldDesc)) {
+         fnWalkFieldTree(subFieldDesc);
+      }
+   };
+
+   fnWalkFieldTree(desc.GetFieldZero());
+
+   // Add the streamer info records from unsplit fields: because of runtime polymorphism we may need to add additional
+   // types not covered by the type names stored in the field headers
+   for (const auto &extraTypeInfo : desc.GetExtraTypeInfoIterable()) {
+      if (extraTypeInfo.GetContentId() != EExtraTypeInfoIds::kStreamerInfo)
+         continue;
+      streamerInfoMap.merge(RNTupleSerializer::DeserializeStreamerInfos(extraTypeInfo.GetContent()).Unwrap());
```

Comment:
It 'seems' that for the streamerInfo of the unpslit branches we serialized then deserialize the StreamerInfo in order to merge them.  If I am correct, can we avoid this (spurrious?) round trip? If I am not, what am I missing?

Reply:
You're right, we serialize and deserialize right away. We could optimize this but I think we should favor code simplicity here. I added a comment in the follow-up PR #16379